src/manifest.rs

```diff
@@ -0,0 +1,442 @@
+        {
+            let core = self.core.lock().unwrap();
+            if core.file.is_none() {
+                return Ok(());
+            }
+        }
+
+        // we could drop the lock during encoding
+        let changes = ManifestChangeSet {
+            changes: changes_param,
+        };
+        let mut buf = BytesMut::new();
+        changes.encode(&mut buf)?;
+
+        let mut core = self.core.lock().unwrap();
+
+        apply_change_set(&mut core.manifest, &changes)?;
+
+        if core.manifest.deletions > self.deletions_rewrite_threshold
+            && core.manifest.deletions
+                > MANIFEST_DELETIONS_RATIO * (core.manifest.creations - core.manifest.deletions)
+        {
+            core.rewrite(&self.directory)?;
+        } else {
+            let mut len_crc_buf = vec![0; 8];
+            (&mut len_crc_buf[..4]).put_u32(buf.len() as u32);
+            (&mut len_crc_buf[4..]).put_u32(crc32::checksum_castagnoli(&buf));
+            len_crc_buf.extend_from_slice(&buf);
+            write_with_length_check!(core.file.as_mut().unwrap(), len_crc_buf);
+        }
+
+        core.file.as_mut().unwrap().sync_all()?;
+        Ok(())
+    }
+
+    pub fn manifest_cloned(&self) -> Manifest {
+        self.core.lock().unwrap().manifest.clone()
+    }
+}
+
+fn apply_change_set(build: &mut Manifest, change_set: &ManifestChangeSet) -> Result<()> {
+    for change in &change_set.changes {
+        apply_manifest_change(build, change)?;
+    }
+    Ok(())
+}
+
+fn apply_manifest_change(build: &mut Manifest, tc: &ManifestChange) -> Result<()> {
+    match ManifestChangeOp::from_i32(tc.op).unwrap() {
+        ManifestChangeOp::Create => {
+            if build.tables.contains_key(&tc.id) {
+                return Err(Error::CustomError(format!(
+                    "manifest invalid, table {} exists",
+                    tc.id
+                )));
+            }
+            build.tables.insert(
+                tc.id,
+                TableManifest {
+                    level: tc.level as u8,
+                    key_id: tc.key_id,
+                },
+            );
+            while build.levels.len() <= tc.level as usize {
+                build.levels.push(LevelManifest::default());
+            }
+            build.levels[tc.level as usize].tables.insert(tc.id);
+            build.creations += 1;
+        }
+        ManifestChangeOp::Delete => {
+            if let Some(tm) = build.tables.get(&tc.id) {
+                assert!(build.levels[tm.level as usize].tables.remove(&tc.id));
+                build.tables.remove(&tc.id);
+                build.deletions += 1;
+            } else {
+                return Err(Error::CustomError(format!(
+                    "manifest invalid, removing non-existing table {}",
+                    tc.id
+                )));
+            }
+        }
+    }
+    Ok(())
+}
+
+pub fn new_create_change(id: u64, level: usize, key_id: u64) -> ManifestChange {
```

Comment:
I prefer to define this function on `ManifestChange`. e.g., ManifestChange::new_create, ManifestChange::new_delete

Reply:
`ManifestChange` is defined in external crate by `prost`, maybe we could not implement method for it directly?

Reply:
Get it, that's true 🤣